drivers/sensor/qdec_nrfx/qdec_nrfx.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sensor.h>
+#include <spinlock.h>
+
+#include <nrfx_qdec.h>
+#include <hal/nrf_gpio.h>
+
+#define SYS_LOG_DOMAIN "qdec_nrfx"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_QDEC_NRFX_LEVEL
+#include <logging/sys_log.h>
+
+
+#define FULL_ANGLE 360
+
+/* limit range to avoid overflow when converting steps to degrees */
+#define ACC_MAX (INT_MAX / FULL_ANGLE)
+#define ACC_MIN (INT_MIN / FULL_ANGLE)
+
+
+struct qdec_nrfx_data {
+	struct k_spinlock        lock;
+	s32_t                    acc;
+	sensor_trigger_handler_t data_ready_handler;
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	u32_t                    pm_state;
+#endif
+};
+
+
+static struct qdec_nrfx_data qdec_nrfx_data;
+
+DEVICE_DECLARE(qdec_nrfx);
+
+
+static void accumulate(struct qdec_nrfx_data *data, int16_t acc)
+{
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
```

Comment:
Why are you using a spinlock?

Reply:
`data->acc` is accessed from two contexts

Reply:
Why not a semaphore, which is used in most other places?

Reply:
@pdunaj agree with @MaureenHelm here. We are not getting the kernel ready for SMP yet. At this point a semaphore is the lightweight locking mechanism that most of all  use for this sort of synchronization.

Reply:
@carlescufi - semaphore may sleep and this is called from the interrupt.
I could disable irqs but I see no point of it if we have splinlocks and can use them directly.

Reply:
Well, I do not even know if these are well-implemented on Arm yet. No other driver uses them as far as I know, so I would try to be conservative and simply use irq_lock/unlock

Reply:
Changed

If spinlocks are considered unsafe documentation should clearly say so or they should be removed altogether.
In our case I don't think they are however as for non smp cases they will be converted to irq_lock.

Reply:
@pdunaj I do agree with you, but I believe it's premature to use spinlocks at this point.